elements/pl-big-o-input/big_o_utils.py

```diff
@@ -0,0 +1,252 @@
+    )
+
+    if sym_true.equals(sym_sub):
+        return (1, CORRECT_COMPLEX_FEEDBACK)
+    elif sym_sub.equals(sympy.sympify(0)):
+        return (0, INCORRECT_FEEDBACK)
+
+    if sympy.limit(sym_sub, sympy.Symbol(variables[0]), sympy.oo) < sympy.sympify(0):
+        return (0, NEGATIVE_FEEDBACK)
+
+    L = sympy.limit(sym_true / sym_sub, sympy.Symbol(variables[0]), sympy.oo)
+
+    if L < sympy.sympify(0):
+        return (0, NEGATIVE_FEEDBACK)
+    elif L == sympy.oo:
+        return (0.25, TOO_LOOSE_FEEDBACK)
+    elif L == sympy.sympify(0):
+        return (0, INCORRECT_FEEDBACK)
+    elif L == sympy.sympify(1):
+        return (0.5, LOWER_ORDER_TERMS_FEEDBACK)
+
+    return (0.5, CONSTANT_FACTORS_FEEDBACK)
+
+
+# Start of generic utilities. Move these to prairielearn.py as-needed
+
+
+class DisplayType(Enum):
+    INLINE = "inline"
+    BLOCK = "block"
+
+
+def grade_question_parameterized(
+    data: pl.QuestionData,
+    question_name: str,
+    grade_function: Callable[[Any], Tuple[Union[bool, float], Optional[str]]],
+    weight: int = 1,
+) -> None:
+    """
+    Grade question question_name. grade_function should take in a single parameter
+    (which will be the submitted answer) and return a 2-tuple:
+        - The first element of the 2-tuple should either be:
+            - a boolean indicating whether the question should be marked correct
+            - a partial score between 0 and 1, inclusive
+        - The second element of the 2-tuple should either be:
+            - a string containing feedback
+            - None, if there is no feedback (usually this should only occur if the answer is correct)
+    """
+
+    # Create the data dictionary at first
+    data["partial_scores"][question_name] = {"score": 0.0, "weight": weight}
+
+    if question_name not in data["submitted_answers"]:
+        data["format_errors"][question_name] = "No answer was submitted"
+        return
+
+    submitted_answer = data["submitted_answers"][question_name]
+
+    # Run passed-in grading function
+    result, feedback_content = grade_function(submitted_answer)
+
+    # Try converting partial score
+    if isinstance(result, bool):
+        partial_score = 1.0 if result else 0.0
+    elif isinstance(result, (float, int)):
+        assert 0.0 <= result <= 1.0
+        partial_score = result
+    else:
+        assert_never(result)
+
+    # Set corresponding partial score and feedback
+    data["partial_scores"][question_name]["score"] = partial_score
+
+    if feedback_content:
+        data["partial_scores"][question_name]["feedback"] = feedback_content
+
+
+def determine_score_params(score: float) -> Tuple[str, Union[bool, float]]:
+    """Determine score params taken from data dict"""
+
+    if score >= 1:
+        return ("correct", True)
+    elif score > 0:
+        return ("partial", math.floor(score * 100))
+
+    return ("incorrect", True)
+
+
+EnumT = TypeVar("EnumT", bound=Enum)
+
+
+def get_enum_attrib(
```

Comment:
Consider adding a test for this function, although I don't think that's super-critical until we lift it into `prairielearn.py`.

Reply:
I think I'd prefer to wait until this gets lifted.